pkg/ext/stores/tokens/tokens.go

```diff
@@ -0,0 +1,1459 @@
+	// GenerateName is not set. as it squashes the name in that case.
+	if token.Name != "" {
+		// reject creation of a token which already exists
+		currentSecret, err := t.secretCache.Get(TokenNamespace, token.Name)
+		if err == nil && currentSecret != nil {
+			return nil, apierrors.NewAlreadyExists(group, token.Name)
+		}
+	}
+
+	// reject user-provided token value, or hash
+	if token.Status.TokenValue != "" {
+		return nil, apierrors.NewBadRequest("User provided token value is not permitted")
+	}
+	if token.Status.TokenHash != "" {
+		return nil, apierrors.NewBadRequest("User provided token hash is not permitted")
+	}
+
+	// Get derived User information for the token.
+	//
+	// NOTE: The creation process is not given `AuthProvider`, `User-` and `GroupPrincipals`.
+	// ..... This information has to be retrieved from somewhere else in the system.
+	// ..... This is in contrast to the Norman tokens who get this information either
+	// ..... as part of the Login process, or by copying the information out of the
+	// ..... base token the new one is derived from. None of that is possible here.
+	//
+	// The token store gets the necessary information (auth provider and
+	// principal id) from the extras of the user stored in the context.
+	//
+	// The `UserPrincipal` data is filled in part from the above, and in
+	// part from the associated `User`s fields.
+	//
+	// `ProviderInfo` is not supported. Norman tokens have it as legacy fallback to hold the
+	// `access_token` data managed by OIDC-based auth providers. The actual primary storage for
+	// this is actually a regular k8s Secret associated with the User.
+
+	user, err := t.userClient.Get(token.Spec.UserID)
+	if err != nil {
+		return nil, apierrors.NewInternalError(fmt.Errorf("failed to retrieve user %s: %w",
+			token.Spec.UserID, err))
+	}
+
+	// Reject operation if the user is disabled.
+	if user.Enabled != nil && !*user.Enabled {
+		return nil, apierrors.NewBadRequest("operation references a disabled user")
+	}
+
+	// Retrieve auth provider and principal id for the user associated with
+	// the authenticated token of the request.
+
+	authProvider, principalID, err := t.auth.ProviderAndPrincipal(ctx, t)
+	if err != nil {
+		return nil, apierrors.NewInternalError(err)
+	}
+
+	// Auto-fill the spec with the principal id, or ensure that the
+	// specified id matches the request. Same as we do for the user id.
+	if token.Spec.PrincipalID == "" {
+		token.Spec.PrincipalID = principalID
+	} else if token.Spec.PrincipalID != principalID {
+		return nil, apierrors.NewBadRequest("unable to create token for other principal id")
+	}
+
+	// Generate secret and its hash
+
+	tokenValue, hashedValue, err := t.hasher.MakeAndHashSecret()
+	if err != nil {
+		return nil, err
+	}
+
+	token.Status.TokenHash = hashedValue
+	token.Status.AuthProvider = authProvider
+	token.Status.DisplayName = user.DisplayName
+	token.Status.UserName = user.Username
```

Comment:
As discussed this information cannot be taken from the user. It should be set verbatim form the authentication token.

Reply:
done in the rewrite to user principal in spec